internal/test/writer.go

```diff
@@ -0,0 +1,27 @@
+package test
+
+import (
+	"io"
+)
+
+// WriterWithHook is an io.Writer that calls a hook function after every write.
+// This is useful for testing.
+type WriterWithHook struct {
+	actualWriter io.Writer
+	hook         func([]byte)
+}
+
+// Write writes p to the actual writer and then calls the hook function.
+func (w *WriterWithHook) Write(p []byte) (n int, err error) {
+	defer w.hook(p)
+	return w.actualWriter.Write(p)
+}
+
+var _ io.Writer = (*WriterWithHook)(nil)
+
+// NewWriterWithHook returns a new WriterWithHook that still writes to the actualWriter
+// but also calls the hook function after every write.
+// The hook function is useful for testing, or waiting for a write to complete.
```

Comment:
And then this one is basically the same as the one above, but with a bit more information. Maybe we can move some things around?